diff --git a/server/server/settings.py b/server/server/settings.py
index 1a8f245..e0bb9e2 100644
--- a/server/server/settings.py
+++ b/server/server/settings.py
@@ -288,11 +288,11 @@ REDIS_PORT = 6379
 REDIS_DB = 0
 
 # Celery configuration
-# USE_CELERY = True
-# CELERY_ACCEPT_CONTENT = ['json']
-# CELERY_TASK_SERIALIZER = 'json'
-# CELERY_RESULT_SERIALIZER = 'json'
-# CELERY_TRIAGE_MEMCACHE_ENTRIES = 100
+USE_CELERY = True
+CELERY_ACCEPT_CONTENT = ['json']
+CELERY_TASK_SERIALIZER = 'json'
+CELERY_RESULT_SERIALIZER = 'json'
+CELERY_TRIAGE_MEMCACHE_ENTRIES = 100
 # CELERY_TASK_ROUTES = {
 #     'crashmanager.cron.*': {'queue': 'cron'},
 #     'ec2spotmanager.cron.*': {'queue': 'cron'},
